### ISSUE TITLE
CV12 raises AssertionError on valid SQL when a clause follows the WHERE clause

### ISSUE THREAD

**Issue by devYRPauli** [sqlfluff/sqlfluff#8430]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Rule CV12 raises `AssertionError` on valid ANSI SQL. The linter reports it as an internal error and asks the user to open an issue. The default rule set triggers it, so no `--rules` flag is needed.

The trigger is a join condition in a WHERE clause with any clause after it. `ORDER BY`, `GROUP BY` and `LIMIT` all reproduce it.

### Expected Behaviour

CV12 reports the join condition and, under `fix`, moves it to an `ON` clause. The clause after the WHERE clause stays where it is.

### Observed Behaviour

```
$ printf 'SELECT foo.a FROM foo INNER JOIN bar WHERE foo.x = bar.y ORDER BY foo.a;\n' \
  | sqlfluff lint --dialect ansi -
CRITICAL   [CV12] Applying rule CV12 to 'stdin' threw an Exception:
Traceback (most recent call last):
  File ".../sqlfluff/core/rules/base.py", line 619, in crawl
    res = self._eval(context=context)
  File ".../sqlfluff/rules/convention/CV12.py", line 65, in _eval
    return [lint_result for lint_result in self._eval_gen(context)]
  File ".../sqlfluff/rules/convention/CV12.py", line 290, in _eval_gen
    assert select_statement.segments[-1].is_type("where_clause")
AssertionError
== [stdin] FAIL
L:   1 | P:   1 | CV12 | Unexpected exception: ;
```

`src/sqlfluff/rules/convention/CV12.py:289-297` reaches the WHERE clause by position:

```python
else:
    assert select_statement.segments[-1].is_type("where_clause")
    assert select_statement.segments[-2].is_type("whitespace", "newline")
    yield LintResult(
        anchor=where_clause,
        fixes=[
            LintFix.delete(select_statement.segments[-2]),
            LintFix.delete(select_statement.segments[-1]),
        ],
    )
```

This branch runs when the whole WHERE clause moves to the join, so the rule deletes it. The WHERE clause is the last child of the select statement only when nothing follows it. `ORDER BY`, `GROUP BY` and `LIMIT` each add a later child.

The other arm of the same `if/else` at `CV12.py:278-288` reaches its target by lookup (`where_clause.get_child("expression")`), not by position.

### How to reproduce

```
printf 'SELECT foo.a FROM foo INNER JOIN bar WHERE foo.x = bar.y ORDER BY foo.a;\n' | sqlfluff lint --dialect ansi -
printf 'SELECT foo.a FROM foo INNER JOIN bar WHERE foo.x = bar.y GROUP BY foo.a;\n' | sqlfluff lint --dialect ansi -
printf 'SELECT foo.a FROM foo INNER JOIN bar WHERE foo.x = bar.y LIMIT 1;\n'         | sqlfluff lint --dialect ansi -
```

Each of the three crashes. The same query without the trailing clause is the control and gives two normal CV12 violations:

```
$ printf 'SELECT foo.a FROM foo INNER JOIN bar WHERE foo.x = bar.y;\n' | sqlfluff lint --dialect ansi -
== [stdin] FAIL
L:   1 | P:  23 | CV12 | Use `JOIN ... ON ...` instead of `WHERE ...` for join
                       | conditions. [convention.join_condition]
```

### Dialect

ansi. The branch is dialect independent.

### Version

```
sqlfluff, version 4.3.0
Python 3.12.12
```

Reproduced on `ddd83f1d7`.

### Configuration

None. The command line above carries the whole configuration.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
